### PR TITLE
Project spec refactor (part 2... )

### PR DESCRIPTION
**Pull request by jmamma**
Part one unintentionally closed after merge.
 
https://github.com/jmamma/MIDICtrl20_MegaCommand/pull/127

--

Things left to do:

- Group Save (add other groups for Tempo and Routing)
- Add MDPolyTrack (for saving poly track select to grid)
- Add LFOTrack (for saving LFO to grid)
- A4 Sound
- Project conversion

### CODE REVIEW

**Comment by jmamma** [comment]
@yatli 

javascript:Array.from(document.getElementsByClassName("ajax-pagination-btn")).forEach(l => l.click());

https://github.com/sindresorhus/refined-github/issues/781

edit: does not expand everything :(

**Comment by yatli** [comment]
> javascript:Array.from(document.getElementsByClassName("ajax-pagination-btn")).forEach(l => l.click());

This has to be called iteratively...

**Comment by yatli** [comment]
Back then the forums have paginators. 🤷‍♂️

**Comment by yatli** [comment]
MNM disconnected.
A4 connected.
Ready for another round of reveng.

**Comment by jmamma** [comment]
PolyGroups: @yatli did you still want to make chromatic mode mutli-timbral. 

Related to this, one of our users has asked for the ability to record in to the sequencer from an external MIDI drum pad, to add this functionality will require isolating control channels for MD drum and MD chromatic.

**Comment by yatli** [comment]
Everything running cool?

**Comment by jmamma** [comment]
My grandfather just passed away.

No coding for a little while, but i'll be back in a week or so 👍

**Comment by yatli** [comment]
I'm sorry to hear that. RIP.

**Comment by yatli** [comment]
`git rev-list --left-right --count dev...ext_conv`
128 0

Dev is well ahead of extconv now.
Merge this PR and start part 3?

File: avr/cores/megacommand/MCL/DeviceTrack.h [hunk 1]
```diff
@@ -86,9 +86,9 @@ class DeviceTrack : public GridTrack {
     if (!that->GridTrack::load_from_mem(col)) {
       return nullptr;
     }
-    if (that->active != track_type) {
+   /* if (that->active != track_type) {
```
Comment (yatli) [inline avr/cores/megacommand/MCL/DeviceTrack.h]: sigh. I should've told myself that a nullptr will backfire, someday.

Comment (jmamma) [reply avr/cores/megacommand/MCL/DeviceTrack.h]: reverted this one.

this fix was elsewhere: 214ace599a9e401ebc7199d2bc0a094befae89ad

**Review by jmamma — COMMENTED** [review]
@yatli Can I get an extra pair of 👀 on this one.

We will need to role out the same type of change for ExtSeqTrack.

Motivation: looks like defer_rendering branch is unstable. We need to optimise event event scanning in seq().

**Comment by yatli** [comment]
> @yatli Can I get an extra pair of 👀 on this one.

which one? you mean this entire set?

**Comment by yatli** [comment]
going over the files.

File: avr/cores/megacommand/MCL/A4SeqTrack.h [hunk 2]
```diff
@@ -7,12 +7,8 @@
 class A4SeqTrack : public ExtSeqTrack {
 
 public:
-  A4SeqTrack(Encoder *e1 = NULL,
-               Encoder *e2 = NULL, Encoder *e3 = NULL, Encoder *e4 = NULL)
-      : ExtSeqTrack(e1, e2, e3, e4) {}
-  bool handleEvent(gui_event_t *event);
-  void display();
-  void setup();
+  A4SeqTrack()
```
Comment (yatli) [inline avr/cores/megacommand/MCL/A4SeqTrack.h:10]: hmm why did we put the encoders here?

**Comment by jmamma** [comment]
too late to the party now.

I made a number of changes to remove locate() and get_lockidx() from Seq() routines.
By dynamically calculating the event_idx on each step.

I tested the changes earlier today. seem to be working.

**Comment by yatli** [comment]
uh oh :)
the @'s are a bit scattered and I couldn't keep up

File: avr/cores/megacommand/MCL/A4SeqTrack.h (hunk 2 above)
Comment (jmamma) [reply avr/cores/megacommand/MCL/A4SeqTrack.h:10]: No idea. I think the history here is messed up. Showing very old changes.

File: avr/cores/megacommand/MCL/ExtSeqTrack.cpp [hunk 3]
```diff
@@ -16,57 +16,239 @@ void ExtSeqTrack::set_speed(uint8_t _speed) {
 
 void ExtSeqTrack::set_length(uint8_t len) {
   length = len;
-  if (step_count >= length) {
+  while (step_count >= length && length > 0) {
```
Comment (yatli) [inline avr/cores/megacommand/MCL/ExtSeqTrack.cpp]: why would a negative value be passed in here?

File: avr/cores/megacommand/MCL/A4SeqTrack.h (hunk 2 above)
Comment (yatli) [reply avr/cores/megacommand/MCL/A4SeqTrack.h:10]: `ext_conv` was quite a while ago

File: avr/cores/megacommand/MCL/ExtSeqTrack.cpp (hunk 3 above)
Comment (jmamma) [reply avr/cores/megacommand/MCL/ExtSeqTrack.cpp]: prevent lockup if length == 0.

**Comment by yatli** [comment]
@jmamma interestingly ba2510b saves 3190 bytes

File: avr/cores/megacommand/MD/MDMessages.cpp [hunk 4]
```diff
@@ -153,7 +153,7 @@ uint16_t MDGlobal::toSysex(ElektronDataToSysexEncoder *encoder) {
   encoder->begin();
   encoder->pack(machinedrum_sysex_hdr, sizeof(machinedrum_sysex_hdr));
   encoder->pack8(MD_GLOBAL_MESSAGE_ID);
-  encoder->pack8(MDX_KIT_VERSION); // version
+  encoder->pack8(0x05); // version
```
Comment (yatli) [inline avr/cores/megacommand/MD/MDMessages.cpp]: 05 or 64?

Comment (jmamma) [reply avr/cores/megacommand/MD/MDMessages.cpp]: 0x05. global version hasn't changed. this was a revert.

**Pull request closed without merging**